pkg/controller/pxcrestore/controller.go

```diff
@@ -102,305 +105,317 @@ func (r *ReconcilePerconaXtraDBClusterRestore) Reconcile(ctx context.Context, re
-		if err != nil {
-			return reconcile.Result{}, err
-		}
-
-		annotations := cr.GetAnnotations()
-		_, unsafePITR := annotations[api.AnnotationUnsafePITR]
-		cond := meta.FindStatusCondition(bcp.Status.Conditions, api.BackupConditionPITRReady)
-		if cond != nil && cond.Status == metav1.ConditionFalse && !unsafePITR {
-			msg := fmt.Sprintf("Backup doesn't guarantee consistent recovery with PITR. Annotate PerconaXtraDBClusterRestore with %s to force it.", api.AnnotationUnsafePITR)
-			err = errors.New(msg)
-			return reconcile.Result{}, nil
-		}
-	}
 
-	err = r.validate(ctx, cr, bcp, cluster)
-	if err != nil {
-		err = errors.Wrap(err, "failed to validate restore job")
+	if err := cr.CheckNsetDefaults(); err != nil {
+		statusState = api.RestoreFailed
+		statusMsg = err.Error()
 		return rr, err
 	}
 
-	log.Info("stopping cluster", "cluster", cr.Spec.PXCCluster)
-	err = r.setStatus(cr, api.RestoreStopCluster, "")
-	if err != nil {
-		err = errors.Wrap(err, "set status")
-		return rr, err
-	}
-	err = k8s.PauseClusterWithWait(ctx, r.client, cluster, true)
-	if err != nil {
-		err = errors.Wrapf(err, "stop cluster %s", cluster.Name)
-		return rr, err
+	cluster := new(api.PerconaXtraDBCluster)
+	if err := r.client.Get(ctx, types.NamespacedName{Name: cr.Spec.PXCCluster, Namespace: cr.Namespace}, cluster); err != nil {
+		if k8serrors.IsNotFound(err) {
+			statusState = api.RestoreFailed
+			statusMsg = err.Error()
+		}
+		return rr, errors.Wrapf(err, "get cluster %s", cr.Spec.PXCCluster)
 	}
 
-	log.Info("starting restore", "cluster", cr.Spec.PXCCluster, "backup", cr.Spec.BackupName)
-	err = r.setStatus(cr, api.RestoreRestore, "")
-	if err != nil {
-		err = errors.Wrap(err, "set status")
-		return rr, err
+	if err := cluster.CheckNSetDefaults(r.serverVersion, log); err != nil {
+		statusState = api.RestoreFailed
+		statusMsg = err.Error()
+		return rr, errors.Wrap(err, "wrong PXC options")
 	}
 
-	err = r.restore(ctx, cr, bcp, cluster)
+	bcp, err := getBackup(ctx, r.client, cr)
 	if err != nil {
-		err = errors.Wrap(err, "run restore")
-		return rr, err
+		statusState = api.RestoreFailed
+		statusMsg = err.Error()
+		return rr, errors.Wrap(err, "get backup")
 	}
 
-	log.Info("starting cluster", "cluster", cr.Spec.PXCCluster)
-	err = r.setStatus(cr, api.RestoreStartCluster, "")
+	restorer, err := r.getRestorer(ctx, cr, bcp, cluster)
 	if err != nil {
-		err = errors.Wrap(err, "set status")
-		return rr, err
+		return rr, errors.Wrap(err, "failed to get restorer")
 	}
 
-	if cr.Spec.PITR != nil {
-		oldSize := cluster.Spec.PXC.Size
-		oldUnsafePXCSize := cluster.Spec.Unsafe.PXCSize
-		oldUnsafeProxySize := cluster.Spec.Unsafe.ProxySize
-
-		var oldProxySQLSize int32
-		if cluster.Spec.ProxySQL != nil {
-			oldProxySQLSize = cluster.Spec.ProxySQL.Size
-		}
-		var oldHAProxySize int32
-		if cluster.Spec.HAProxy != nil {
-			oldHAProxySize = cluster.Spec.HAProxy.Size
+	switch statusState {
```

Comment:
maybe we should move code for each case to a separate function? this would make code more readable

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/1610/commits/fa207d39adde7d794affaf7d84536f1fbbbcf274